File: 81-SearchInRotatedSortedArrayIi/81-SearchInRotatedSortedArrayIi.py
# Last updated: 7/12/2026, 6:23:46 PM
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        s = 0
        e = len(nums) - 1

        while s <= e:
            mid =  (s + e) // 2

            if nums[mid] == target: return True

            if nums[mid] == nums[s] and nums[mid] == nums[e]:
                s += 1
                e -= 1
                continue

            if nums[s] <= nums[mid]:
                if nums[s] <= target and target <= nums[mid]:
                    e = mid - 1
                else:
                    s = mid + 1
            else:
                if nums[mid] <= target and target <= nums[e]:
                    s = mid + 1
                else:
                    e = mid - 1
            
        return False

File: 81-SearchInRotatedSortedArrayIi/81-SearchInRotatedSortedArrayIi.java
// Last updated: 7/12/2026, 6:23:49 PM
class Solution {
	public static boolean search(int[] nums, int target) {
        
        int s = 0;
        int e = nums.length -1;
        
        while(s<=e){
            int mid = s + (e - s) / 2;
            
            if(nums[mid] == target) return true;
            
            // if left side is sorted
            if (nums[mid] > nums[s]) {
                if(target >= nums[s] && target <= nums[mid]){
                    e = mid;
                }else s = mid + 1;
                
            }
            
            // if right side is sorted
            else if(nums[mid] < nums[s]){
                if(target >= nums[mid] && target <= nums[e]){
                    s = mid;
                } else e = mid - 1;
            } else {
                // if nums[s] == nums[e] == nums[mid]
                s++;
            }
        }
        
        return false;
        
        
    }
}